operation/fhir-operation-validate/src/main/java/org/linuxforhealth/fhir/operation/validate/ValidateOperation.java

```diff
@@ -236,53 +244,100 @@ protected FHIROperationException buildExceptionWithIssue(IssueSeverity severity,
+            if (operationContext != null && operationContext.getType() == FHIROperationContext.Type.INSTANCE
+                    && modeType == ModeType.PROFILE) {
+                resource = resourceHelper.doRead(resourceType, logicalId).getResource();
+                if (resource == null) {
+                    throw buildExceptionWithIssue(IssueSeverity.ERROR, IssueType.NOT_FOUND,
+                            resourceType + " with id '" + logicalId + "' does not exist", null);
+                }
+            }
+        } else {
             resource = resourceParameter.getResource();
         }
+
         return resource;
     }
-    
+
+    /**
+     * When $validate is invoked at the instance level, we need to perform some additional validations:
+     * <ul>
+     * <li>if mode = create or update, check that the id of the passed resource matches the id in the URL
+     * <li>if mode = create, check if the resource already exists.
+     * <li>if mode = create or mode = update and the resource doesn't exist, check that "create-on-update" is enabled
+     * </ul>
+     * @param operationContext - the FHIROperationContext associated with the request
+     * @param logicalId - the logical id of the FHIR resource
+     * @param resourceHelper - Resource operation provider for loading related Library resources
+     * @param resource - the resource to be validated
+     * @param resourceType  - a valid resource type string
+     * @param modeType - a valid resource validation mode code
+     * @return List<Issue> - a non-null, possibly-empty list of issues that indicate why the resource is not valid for the passed modeType
+     */
+    private List<Issue> validateInstanceLevelInvoke(FHIROperationContext operationContext, String logicalId, FHIRResourceHelpers resourceHelper,
+            Resource resource, String resourceType, ModeType modeType) throws Exception, FHIROperationException {
+        List<Issue> issues = new ArrayList<>();
+
+        if (modeType == ModeType.CREATE || modeType == ModeType.UPDATE) {
+            // if mode = create or update, check that the id of the passed resource matches the id in the URL
+            validateResourceId(logicalId, resource, issues);
+
+            boolean resourceExists = resourceHelper.doRead(resourceType, logicalId).getResource() != null;
+
+            // if mode = create, check if the resource already exists
+            if (modeType == ModeType.CREATE && resourceExists) {
+                issues.add(FHIRUtil.buildOperationOutcomeIssue(IssueSeverity.ERROR, IssueType.INVALID,
+                        resourceType + " with id '" + logicalId + "' already exists"));
+            }
+
+            // if mode = create or mode = update and the resource doesn't exist yet,
+            // check if the persistence layer implementation supports update/create mode
+            if (modeType == ModeType.CREATE || !resourceExists) {
```

Comment:
I think the code is right, but the comment could perhaps be clearer.  These are the two cases:
1. mode = create
2. mode = update and !resourceExists

however, since we're already inside of a block that ensures mode is either create or update, its reduces to just `(modeType == ModeType.CREATE || !resourceExists)`

Reply:
i added an extra if in the comment to hopefully make it a bit more clear

Reply:
sure. thank you